MediaSession artwork only uses the first item in the artwork array

jyavenard opened the pull request:
#### cbbcae6e1ae3ca3c2b11a499fa133b07746e2820
<pre>
MediaSession artwork only uses the first item in the artwork array
<a href="https://bugs.webkit.org/show_bug.cgi?id=276351">https://bugs.webkit.org/show_bug.cgi?id=276351</a>
<a href="https://rdar.apple.com/81160539">rdar://81160539</a>

Reviewed by Youenn Fablet.

If `sizes` metadata artwork attribute is provided we will use them to determine the &quot;best&quot;
image to select. We do so by giving each resolution a score between 0 and 1. We aim to download
an image that is 512x512 with an aspect ratio of 1.
If the `sizes` attribute contains invalid content or if not provided, we will iterate through
all the available images and download them. We will then either select the one with the highest
score, or the first one found with a resolution greater than 512 pixels.

Test will be added once <a href="https://bugs.webkit.org/show_bug.cgi?id=276133">https://bugs.webkit.org/show_bug.cgi?id=276133</a> lands as it adds some
internal APIs to check which URL was selected with Now Playing.
For now, manually tested.

* Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
(WebCore::MediaMetadata::imageDimensionsScore const):
(WebCore::MediaMetadata::refreshArtworkImage):
(WebCore::MediaMetadata::tryNextArtworkImage):
* Source/WebCore/Modules/mediasession/MediaMetadata.h:

Canonical link: <a href="https://commits.webkit.org/280804@main">https://commits.webkit.org/280804@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/515c5077ab9951a4ae1ab8feeb801164d9a058b3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/57692 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37020 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10168 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61314 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8137 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/59820 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44656 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8325 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46746 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5767 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59722 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34717 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49850 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27568 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31488 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7141 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7141 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/53442 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7413 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62995 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1606 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7489 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53961 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1612 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49861 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54076 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12757 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1357 "Passed tests") | | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/32849 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33935 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/35019 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33680 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
```diff
@@ -159,22 +161,110 @@ ExceptionOr<void> MediaMetadata::setArtwork(ScriptExecutionContext& context, Vec
+            return { m_idealSize, m_idealSize }; // We prefer image tagged with "any" size.
+        StringView stringView(sizes);
+        IntSize size;
+        for (auto element : stringView.split(' ')) {
```
```suggestion
        for (auto element : StringView(sizes).split(' ')) {
```

youennf replied:
Also do we handle the case of multiple spaces like `192x192      320x320`?

jyavenard replied:
> Also do we handle the case of multiple spaces like `192x192 320x320`?

Hmm I didn't think of that. If spaces are correct then we take the biggest of the series

youennf commented on Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
```diff
@@ -159,22 +161,110 @@ ExceptionOr<void> MediaMetadata::setArtwork(ScriptExecutionContext& context, Vec
+        IntSize size;
+        for (auto element : stringView.split(' ')) {
+            auto sizeString = element.convertToASCIILowercase();
+            Vector<std::optional<uint32_t>> numbers;
```
```suggestion
            Vector<std::optional<uint32_t>, 2> numbers;
```

jyavenard replied:
removed that code altogether

youennf commented on Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
```diff
@@ -159,22 +161,110 @@ ExceptionOr<void> MediaMetadata::setArtwork(ScriptExecutionContext& context, Vec
+        StringView stringView(sizes);
+        IntSize size;
+        for (auto element : stringView.split(' ')) {
+            auto sizeString = element.convertToASCIILowercase();
```
We can probably use findIgnoringASCIICase("x"_s) to remove this string allocation.

youennf commented on Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
```diff
@@ -159,22 +161,110 @@ ExceptionOr<void> MediaMetadata::setArtwork(ScriptExecutionContext& context, Vec
+    String artworkImageSrc = m_metadata.artwork[index].src;
+
+    m_artworkLoader = makeUnique<ArtworkImageLoader>(*m_session->document(), artworkImageSrc, [this, index, iterate, artworkImageSrc](Image* image) {
+        IntSize size = image && image->data() ? IntSize(image->width(), image->height()) : IntSize();
```
I prefer the old version of checking whether image is null upfront.
The current version is ok I think for that case, but somehow convoluted.

jyavenard replied:
> I prefer the old version of checking whether image is null upfront. The current version is ok I think for that case, but somehow convoluted.

we don't want to exit early when it's null anymore, we want to try the other artwork

youennf replied:
Sure, wouldn't the following just work:
```
if (!image) {
    tryNextArtworkImage(index + 1);
    return;
}
m_artworkImageSrc = artworkImageSrc;
setArtworkImage(image);
metadataUpdated();
// Maybe we should also free m_artworkLoader, need to check how this loader is implemented.
```

youennf commented on Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
```diff
@@ -159,22 +161,110 @@ ExceptionOr<void> MediaMetadata::setArtwork(ScriptExecutionContext& context, Vec
+        return imageDimensionsScore(sizes[index].width(), sizes[index].height());
+    });
+
+    bool iterate = true;
```
s/iterate/shouldIterate/

jyavenard replied:
removed that code

youennf commented on Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
```diff
@@ -159,22 +161,110 @@ ExceptionOr<void> MediaMetadata::setArtwork(ScriptExecutionContext& context, Vec
-        setArtworkImage(image);
-        metadataUpdated();
+
+        if (!iterate || index + 1 < m_metadata.artwork.size())
```
```suggestion
        if (!iterate || ++index < m_metadata.artwork.size())
```

jyavenard replied:
this would require to make the entire lambda mutable

youennf commented on Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
```diff
@@ -159,22 +161,110 @@ ExceptionOr<void> MediaMetadata::setArtwork(ScriptExecutionContext& context, Vec
+            if (numbers.size() != 2 || !numbers[0] || !numbers[1])
+                return { };
+
+            IntSize newSize = { int(*numbers[0]), int(*numbers[1]) };
```
```suggestion
            IntSize newSize { int(*numbers[0]), int(*numbers[1]) };
```

youennf commented on Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
```diff
@@ -159,22 +161,110 @@ ExceptionOr<void> MediaMetadata::setArtwork(ScriptExecutionContext& context, Vec
+    tryNextArtworkImage(bestIndex, iterate);
+}
+
+void MediaMetadata::tryNextArtworkImage(uint32_t index, bool iterate)
```
Would it be simpler to have a Vector<String> of image srcs, sorted by the image score.
Then, start with the first image src (the one with the best score) and only iterate if there is no image?

jyavenard replied:
If we can sort the src, then we only retrieve one image. There's no need to keep sorted vector anymore. 

Otherwise we can't sort until we have downloaded all the images.

jyavenard replied:
yeah, it ended up being nicer that way. and it simplifies the logic as we don't need to distinguish the case where we do need to iterate for those we don't

youennf commented on Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
```diff
@@ -159,22 +161,101 @@ ExceptionOr<void> MediaMetadata::setArtwork(ScriptExecutionContext& context, Vec
+    auto aspectRatioCoefficient = shortEdge / longEdge;
+
+    if (longEdge < m_idealSize)
+        return aspectRatioCoefficient * (0.8 * (longEdge - m_minimumSize) / (m_idealSize - m_minimumSize) + 0.2);
```
Theoretically, it would be possible for m_idealSize and m_minimumSize to be the same.
Given they are static, we should probably rename them as s_...
Or maybe make this method a free function and have const variables for ideal/minimum sizes.

jyavenard replied:
I added a static_assert

youennf commented on Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
```diff
@@ -159,22 +161,101 @@ ExceptionOr<void> MediaMetadata::setArtwork(ScriptExecutionContext& context, Vec
+        return size;
+    });
+
+    m_sortedSrc = { sizes.size(), [&](size_t index) -> Pair {
```
We probably do not need to have i m_sortedSrc  as a class member.
We can pass it to tryNextArtworkImage which would handle it.

youennf commented on Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
```diff
@@ -159,22 +161,101 @@ ExceptionOr<void> MediaMetadata::setArtwork(ScriptExecutionContext& context, Vec
+    m_artworkImageSrc = String();
+    m_artworkImage = nullptr;
+    m_artworkLoader = nullptr;
+    m_sortedSrc.clear();
```
```suggestion
    m_sortedSrces.clear();
```

youennf commented on Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
```diff
@@ -159,22 +161,101 @@ ExceptionOr<void> MediaMetadata::setArtwork(ScriptExecutionContext& context, Vec
+
+    m_artworkLoader = makeUnique<ArtworkImageLoader>(*m_session->document(), artworkImageSrc, [this, index, artworkImageSrc](Image* image) {
+        IntSize size = image && image->data() ? IntSize(image->width(), image->height()) : IntSize();
+        float imageScore = size.isEmpty() ? -1.0 : imageDimensionsScore(size.width(), size.height());
```
```suggestion
        float imageScore = imageDimensionsScore(size.width(), size.height());
```

youennf commented on Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
```diff
@@ -159,22 +161,101 @@ ExceptionOr<void> MediaMetadata::setArtwork(ScriptExecutionContext& context, Vec
+        IntSize size = image && image->data() ? IntSize(image->width(), image->height()) : IntSize();
+        float imageScore = size.isEmpty() ? -1.0 : imageDimensionsScore(size.width(), size.height());
+        if (imageScore >= 0
+            && (!index || (m_artworkImage && (imageDimensionsScore(m_artworkImage->width(), m_artworkImage->height()) < imageScore)))) {
```
Do we even need this now?
Shouldn't we take the first src for which we have a valid Image, given it is sorted?

youennf commented on Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
```diff
@@ -159,22 +161,101 @@ ExceptionOr<void> MediaMetadata::setArtwork(ScriptExecutionContext& context, Vec
-        setArtworkImage(image);
-        metadataUpdated();
+
+        if (index + 1 < m_sortedSrc.size())
```
```suggestion
        if (++index < m_sortedSrc.size())
```

youennf commented on Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
```diff
@@ -159,22 +161,101 @@ ExceptionOr<void> MediaMetadata::setArtwork(ScriptExecutionContext& context, Vec
-        metadataUpdated();
+
+        if (index + 1 < m_sortedSrc.size())
+            tryNextArtworkImage(index + 1);
```
```suggestion
            tryNextArtworkImage(index);
```

youennf commented on Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
```diff
@@ -159,22 +161,101 @@ ExceptionOr<void> MediaMetadata::setArtwork(ScriptExecutionContext& context, Vec
+    tryNextArtworkImage(0);
+}
+
+void MediaMetadata::tryNextArtworkImage(uint32_t index)
```
I would tend to do the following:
```suggestion
void MediaMetadata::tryNextArtworkImage(Vector<Pair>&&, uint32_t index = 0)
```

And have the lambda capture the Vector.
Also `m_sortedSrc`and `m_artworkLoader` are useless once we finished loading. If it was captured in the lambda, it would be freed automatically by freeing `m_artworkLoader`

youennf commented on Source/WebCore/Modules/mediasession/MediaMetadata.h:
```diff
@@ -100,12 +100,22 @@ class MediaMetadata : public RefCounted<MediaMetadata> {
+    void tryNextArtworkImage(uint32_t);
+    float imageDimensionsScore(int, int) const;
+
+    static constexpr int m_minimumSize = 128;
```
```suggestion
    static constexpr int s_minimumSize = 128;
```

youennf commented on Source/WebCore/Modules/mediasession/MediaMetadata.h:
```diff
@@ -100,12 +100,22 @@ class MediaMetadata : public RefCounted<MediaMetadata> {
+    float imageDimensionsScore(int, int) const;
+
+    static constexpr int m_minimumSize = 128;
+    static constexpr int m_idealSize = 512;
```
```suggestion
    static constexpr int s_idealSize = 512;
```

youennf commented on Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
```diff
@@ -159,22 +161,101 @@ ExceptionOr<void> MediaMetadata::setArtwork(ScriptExecutionContext& context, Vec
-        if (!image || !image->data())
+
+    // First look into the artwork's sizes attributes to attempt to determine the best score.
+    Vector<IntSize> sizes(m_metadata.artwork.size(), [&](size_t index) -> IntSize {
```
Would it be possible to directly create m_sortedSrc without creating this sizes vector?

youennf reviewed (approved):
LGTM.
Some further improvements added in comments.

jyavenard committed:
MediaSession artwork only uses the first item in the artwork array
https://bugs.webkit.org/show_bug.cgi?id=276351
rdar://81160539

Reviewed by Youenn Fablet.

If `sizes` metadata artwork attribute is provided we will use them to determine the "best"
image to select. We do so by giving each resolution a score between 0 and 1. We aim to download
an image that is 512x512 with an aspect ratio of 1.
If the `sizes` attribute contains invalid content or if not provided, we will iterate through
all the available images and download them. We will then either select the one with the highest
score, or the first one found with a resolution greater than 512 pixels.

Test will be added once https://bugs.webkit.org/show_bug.cgi?id=276133 lands as it adds some
internal APIs to check which URL was selected with Now Playing.
For now, manually tested.

* Source/WebCore/Modules/mediasession/MediaMetadata.cpp:
(WebCore::MediaMetadata::imageDimensionsScore const):
(WebCore::MediaMetadata::refreshArtworkImage):
(WebCore::MediaMetadata::tryNextArtworkImage):
* Source/WebCore/Modules/mediasession/MediaMetadata.h:

Canonical link: https://commits.webkit.org/280804@main
